### PR TITLE
RIA-2526 downstream services health check

### PR DESCRIPTION
**Pull request by tarun-palisetty**
### JIRA link (if applicable) ###
[RIA-2526](https://tools.hmcts.net/jira/browse/RIA-2526)

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/DownStreamHealthIndicator.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    public DownStreamHealthIndicator(HealthCheckConfiguration healthCheckConfiguration) {
+        this.healthCheckConfiguration = healthCheckConfiguration;
+
+        healthCheckConfiguration.getServices().entrySet().stream()
+            .forEach(s -> {
+                contributors
+                    .put(s.getKey(), new ServiceHealthIndicator(
+                        s.getValue().get("uri"),
+                        s.getValue().get("response"),
+                        new RestTemplate()));
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/DownStreamHealthIndicator.java]: Could you use the RestTemplate already configured in the app and inject instead?  Do we really need to call the `new` method here?

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/ServiceHealthIndicator.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+public class ServiceHealthIndicator implements HealthIndicator {
+
+    private String uri;
+    private String matcher;
+    private RestTemplate restTemplate;
+
+    public ServiceHealthIndicator(String uri, String matcher, RestTemplate restTemplate) {
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/ServiceHealthIndicator.java]: Maybe you can inject restTemplate here?

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/ServiceHealthIndicator.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                    .Builder(Status.UP)
+                    .build();
+            } else {
+
+                return new Health
+                    .Builder(Status.DOWN)
+                    .build();
+            }
+        } catch (RestClientException ex) {
+
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/ServiceHealthIndicator.java]: Put a log entry in here to say a http exception was received including the status code / message?

File: src/test/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/DownStreamHealthIndicatorTest.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+import uk.gov.hmcts.reform.iacaseapi.infrastructure.config.HealthCheckConfiguration;
+
+@RunWith(MockitoJUnitRunner.class)
+@SuppressWarnings("unchecked")
+public class DownStreamHealthIndicatorTest {
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/DownStreamHealthIndicatorTest.java]: This unit test needs more tests.  E.g. you are doing a lot of stuff in the constructor which is not tested at all.  E.g. what if a null list of services is retrieved?

Comment (raghera) [reply src/test/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/DownStreamHealthIndicatorTest.java]: You may also consider checking that you are calling the constructor the correct number of times.  You can do that using PowerMockito.

File: src/test/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/ServiceHealthIndicatorTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        assertEquals(Health.down().build(), serviceHealthIndicator.health());
+    }
+
+    @Test
+    public void health_should_throw_exception_rest_error() {
+        when(restTemplate.getForEntity(uri, String.class)).thenThrow(new RestClientException("Internal server error"));
+
+        assertEquals(Health.down(new RestClientException("Internal server error")).build(), serviceHealthIndicator.health());
+    }
+
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/ServiceHealthIndicatorTest.java]: Can probably add tests around getting an OK status code but something being down in the response body or doesn't match the matcher.

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/DownStreamHealthIndicator.java (hunk 1 above)
Comment (bradsorour) [reply src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/health/DownStreamHealthIndicator.java]: Using restTemplate now, had to amend the RestTemplateConfiguration (change default message converter) in order to function correctly.

**Comment by bradsorour** [comment]
Creating new PR to get pipeline green. This PR continually failing on helm upgrade so unable to get past "AKS deploy - preview" step.

**Pull request closed without merging**